File: Data Structures & Algorithms/three-integer-sum/submission-2.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        ans = {}
        for i in range(len(nums)-2):
            l = i+1
            r = len(nums)-1
            while l<r:
                if nums[i]+nums[l]+nums[r]>0:
                    r-=1
                    while r>=0 and nums[r]==nums[r+1]:
                        r-=1
                elif nums[i]+nums[l]+nums[r]<0:
                    l+=1
                    while l<=len(nums)-1 and nums[l]==nums[l-1]:
                        l+=1
                else:
                    ans[(nums[i],nums[l],nums[r])]=1
                    l+=1
                    r-=1
                    while r>0 and nums[r]==nums[r+1]:
                        r-=1
                    while l<len(nums)-1 and nums[l]==nums[l-1]:
                        l+=1
        return [i for i in ans]

File: Data Structures & Algorithms/house-robber-ii/submission-2.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.size()<=3){
            int mx = 0;
            for(auto i : nums) mx = max(i,mx);
            return mx;
        }

        vector<int> org(nums);

        nums[nums.size()-1] = 0;
        nums[2] += nums[0];
        for(int i = 3 ; i < nums.size() ; i++){
            nums[i] += max(nums[i-3], nums[i-2]);
        }
        int mx = max(nums[nums.size()-1], nums[nums.size()-2]);

        nums = org;
        nums[0] = 0;
        for(int i = 3 ; i < nums.size() ; i++){
            nums[i] += max(nums[i-3], nums[i-2]);
        }
        mx = max(nums[nums.size()-1],mx);
        mx = max(nums[nums.size()-2],mx);

        return mx;
    }
};








